Transformations/Transform/Transform.cpp

```diff
@@ -0,0 +1,340 @@
+        double angle = cv::norm(rotvec);
+        cv::Mat axis = rotvec / cv::norm(rotvec);
+
+        return std::make_tuple(angle, axis);
+    }
+
+    cv::Mat rotvec2quat(cv::Mat rotvec, double angle = nan(""))
+    {
+        if(isnan(angle))
+        {
+            angle = cv::norm(rotvec);
+        }
+
+        cv::Mat quat(4, 1, CV_64FC1, cv::Scalar(0));
+        quat.at<double>(0) = cos(0.5 * angle);
+        if(quat.at<double>(0) == 0)
+        {
+            throw "cos(0.5 * angle) = 0";
+        }
+        auto a = sqrt((1 - pow(quat.at<double>(0), 2)) / pow(angle, 2));
+        for(int i = 0; i < 3; i++)
+        {
+            quat.at<double>(i + 1) = a * rotvec.at<double>(i);
+        }
+        quat = quat / SIGN(quat.at<double>(0));
+
+        return quat;
+    }
+
+    cv::Mat quat2rotvec(cv::Mat quat)
+    {
+        cv::Mat rotvec(3, 1, CV_64FC1, cv::Scalar(0));
+        auto angle = acos(quat.at<double>(0)) * 2;
+        auto a = sqrt((1 - pow(quat.at<double>(0), 2)));
+        for(int i = 0; i < 3; i++)
+        {
+            rotvec.at<double>(i) = angle * quat.at<double>(i + 1) / a;
+        }
+
+        return rotvec;
+    }
+
+    cv::Mat angleaxis2rotvec(double angle, cv::Mat axis)
+    {
+        return axis * angle;
+    }
+
+    std::tuple<double, cv::Mat> rotvec2angleaxis(cv::Mat rotvec)
+    {
+        cv::Mat axis(3, 1, CV_64FC1, cv::Scalar(0));
+        double angle = cv::norm(rotvec);
+        axis = rotvec / cv::norm(rotvec);
+
+        return std::make_tuple(angle, axis);
+    }
+
+    std::tuple<double, cv::Mat> quat2angleaxis(cv::Mat quat)
+    {
+        cv::Mat axis(3, 1, CV_64FC1, cv::Scalar(0));
+        auto rotvec = quat2rotvec(quat);
+        double angle = cv::norm(rotvec);
+        axis = rotvec / cv::norm(rotvec);
+
+        return std::make_tuple(angle, axis);
+    }
+
+    cv::Mat rotx(double rad)
+    {
+        cv::Mat rotmat(3, 3, CV_64FC1, cv::Scalar(0));
+        rotmat = cv::Mat::eye(3, 3, CV_64FC1);
+        rotmat.at<double>(1, 1) = cos(rad);
+        rotmat.at<double>(1, 2) = -sin(rad);
+        rotmat.at<double>(2, 1) = sin(rad);
+        rotmat.at<double>(2, 2) = cos(rad);
+
+        return rotmat;
+    }
+
+    cv::Mat roty(double rad)
+    {
+        cv::Mat rotmat(3, 3, CV_64FC1, cv::Scalar(0));
+        rotmat = cv::Mat::eye(3, 3, CV_64FC1);
+        rotmat.at<double>(0, 0) = cos(rad);
+        rotmat.at<double>(0, 2) = sin(rad);
+        rotmat.at<double>(2, 0) = -sin(rad);
+        rotmat.at<double>(2, 2) = cos(rad);
+
+        return rotmat;
+    }
+
+    cv::Mat rotz(double rad)
+    {
+        cv::Mat rotmat(3, 3, CV_64FC1, cv::Scalar(0));
+        rotmat = cv::Mat::eye(3, 3, CV_64FC1);
+        rotmat.at<double>(0, 0) = cos(rad);
+        rotmat.at<double>(0, 1) = -sin(rad);
+        rotmat.at<double>(1, 0) = sin(rad);
+        rotmat.at<double>(1, 1) = cos(rad);
+
+        return rotmat;
+    }
+
+    cv::Mat euler2rotmat(cv::Vec3d euler, std::string orientation = "zyx")
```

Comment:
On the KB we call this roll-pitch-yaw instead of Euler. I chose that notation because that is what is common in robotics literature (Robot Modeling and Control by Spong and Robotics, Vision and Control by Corke) for rotations about three distinct axes. Your man from opencv link thinks it is called Tait-Bryan - maybe in some other areas of science it is.

I don't know what is best but my gut feeling is that people from robotics backgrounds will use this the most.

Reply:
Only reason I didn't want to go with RPY is because it may imply order. As we've seen this is a matter of definition. By calling the vector `euler` we avoid that. That being said, I'm not really opposed to changing it.